### PR TITLE
visual test services

### PR DESCRIPTION
**Pull request by muratkeremozcan**

### CODE REVIEW

File: sections/tools/visual-regression-testing.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        cy.getByCy('user-details').percySnapshotElement(
+          'user details with custom selector'
+        )
+      })
+    })
+  }
+)
+```
+
+We see the first one-liner taking a full screen shot, and the subsequent call taking a snapshot of the `user-details` selector. Selector focused snapshots are not natively supported in Percy, therefore [a custom command](https://github.com/muratkeremozcan/react-hooks-in-action-with-cypress/blob/main/cypress/support/commands.js#L112) `percySnapshotElement` is used.
```
Comment (NoriSte) [inline sections/tools/visual-regression-testing.md:74]: **Suggestion**: Adding the custom command above the test or splitting the tests and explanations into two different parts: the first is a classic snapshot, the second is a selector-based snapshot due to Percy not supporting it, etc.

File: sections/tools/visual-regression-testing.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+Similar to Percy, with Applitools our test executes cross browser & viewport, and records the base snapshot.
+
+![Image description](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/uwxxha4w1wno1w076qwb.png)
+
+When / if there is a visual diff, there is a clear indicator in the web interface.
+
+![Image description](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/wrrat01cd06zz134n1l9.png)
+
+Here is a side by side comparison of the Percy vs Applitools code.
```
Comment (NoriSte) [inline sections/tools/visual-regression-testing.md:158]: The comparison is lacking the custom `percySnapshotElement` custom command needed

File: sections/tools/visual-regression-testing.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+![Image description](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/tbpd7fjt2idxkyur24tg.png)
+
+Overall Applitools is strong on configurability while Percy is strong on simplicity. The UX is leaner and easier to use on Percy side, while on Applitools the UX is busier in comparison, but it has improved much over the years. Percy certainly has less code, not having to "open" and "close" eyes and being able to fire off the main command is a big win. For local developer experience, Applitools is the winner; being able to execute the tests with Cypress open mode vs elaborate CLI commands is huge win. Failing an actual visual diff in the test runner, vs the visual failures being only on the web UI in Percy's case, is also a win for Applitools. For CI, not having to configure any yml makes Applitools the winner there as well. Another win is for being able to take snapshots of sub-sections of the UI via selectors; this feature is built-in to Applitools while with Percy it has to be custom command that is not sure to work everywhere in the real world.
```
Comment (NoriSte) [inline sections/tools/visual-regression-testing.md:162]: What about transforming the whole chapter into a comparison table?

**Review by NoriSte — COMMENTED** [review]
**Praise**: Big **THANK YOU** for writing this ❤️ what a thorough guide rich of on-the-field experience!!!

File: sections/tools/visual-regression-testing.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
+# Visual Regression Testing
+
+<br/><br/>
+
+### One Paragraph Explainer
+
+A visual regression test checks what the user will see after any code changes have been executed by comparing screenshots taken before and after code changes. It is the primary way to validate regressions in CSS, but it is also useful for covering viewport & cross browser/device combinations, as well as localization. Think of Jest Snapshots, but instead of comparing the DOM as text, it is an actual screen shot comparison.
+
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:8]: added

File: sections/tools/visual-regression-testing.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+* Record the base snapshot (initial execution).
+* Execute the visual test again, compare to the base snapshot (subsequent executions).
+  * If the new snapshot matches the base snapshot, accept.
+  * Else, set the new base snapshot, or it is a visual regression defect.
+
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:13]: I changed it in way markdown linter accepts.

File: sections/tools/visual-regression-testing.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+The visual regression tools will find any and every pixel diffs. This may get hectic the bigger the snapshot, and as the number of snapshots increase. **The big selling point of visual snapshot services is the AI; the AI is trainable over time and we can train it to ignore petty snapshot diffs we might not care for**. Be mindful that given a snapshot name, we can train the AI to be really bad by accepting any diff. Therefore we need to take care not to thumbs up valid failures. We can reset the training by changing the snapshot name, viewport, or any part of the code. 
+
+**The second selling point of services is how they can address cross viewport and browser concerns with a single test, so that we do not have to repeat the same test with different variants in CI**.
+
+
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:18]: Added a final line for itin the previous section.

File: sections/tools/visual-regression-testing.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+### Cypress examples with Percy and Applitools
+
+> All the code samples can be found in the [full repo](https://github.com/muratkeremozcan/react-hooks-in-action-with-cypress).
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:22]: Added a description of the repo.

File: sections/tools/visual-regression-testing.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+- Non-matching new snapshots prompt a notification on the web interface; we either have to reject or accept this new baseline. If we reject, it is a defect. If we accept we have a new base line and the cycle continues.
+
+Suppose we want to verify the user's avatar. 
+
+![Image description](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/rnl9mszbi1uuwr0ytgaf.png)
+
+#### Percy flow:
+
+* Here is [a simple test](https://github.com/muratkeremozcan/react-hooks-in-action-with-cypress/blob/main/cypress/e2e/ui-integration/user-context-retainment.spec.js) that verifies the avatar.
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:39]: For me, seeing the real code in context is a lot more important than seeing the code sample. The link does not hurt those who think it does not add value.

File: sections/tools/visual-regression-testing.md [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+
+```javascript
+// cypress/e2e/ui-integration/user-context-retainment.spec.js
+describe('User selection retainment between routes',
+  { tags: '@ui-integration' },
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:45]: It's from cypress-grep. With an easy command you can filter & run the tagged tests. 

The repo isn't only for visual tests. I can remove it from this code sample.

[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:45]: Check out https://dev.to/muratkeremozcan/the-32-ways-of-selective-testing-with-cypress-a-unified-concise-approach-to-selective-testing-in-ci-and-local-machines-1c19

File: sections/tools/visual-regression-testing.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      cy.fixture('users').then((users) => {
+        cy.get('.user-picker').select(users[3].name)
+        cy.contains('Users').click()
+
+        cy.wait('@userStub')
+        cy.url().should('contain', '/users')
+        cy.get('.item-header').contains(users[3].name)
+
+        // the visual test
+        cy.percySnapshot('User selection retainment between routes')
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:62]: Got it. Done.

File: sections/tools/visual-regression-testing.md (hunk 1 above)
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:74]: Made 2 tests.

File: sections/tools/visual-regression-testing.md (hunk 2 above)
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:158]: It's on L25 in user-context-retainment.spec.js, the left side.

File: sections/tools/visual-regression-testing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:20]: Added a final line for it in the previous section.

File: README.md [hunk 11]
```diff
@@ -199,6 +199,16 @@ some UI utilities that make easy to use them as a development tool.
 
 <br/><br/>
 
+## ![✔] 6.2 Visual Regression Testing
+
+**TL;DR:** Visual regression tests hard and why we should rely on premium services.
+
+**Otherwise:** Another continuous chore for regressions we do not care about.
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply README.md]: Added.

File: sections/tools/visual-regression-testing.md [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+        // using custom command for css selector focus
+        cy.getByCy('user-details').percySnapshotElement(
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md]: Done.

File: sections/tools/visual-regression-testing.md (hunk 2 above)
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md:176]: updated screenshot

File: sections/tools/visual-regression-testing.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+});
+```
+
+We realize the additional `cy.eyesOpen` and `cy.eyesClose` commands that need to execute in the beginning and end of the test. We also see ` cy.eyesCheckWindow` is very customizable, not needing a custom command as in Percy.
+
+> For details about setting up Applitools, and comparisons to Percy, check out [this blog post](https://dev.to/muratkeremozcan/setup-cypress-applitools-with-github-actions-a-comparison-of-applitools-vs-percy-in-a-mid-size-app-43ij).
+
+Similar to Percy, with Applitools our test executes cross browser & viewport, and records the base snapshot.
+
+![Image description](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/uwxxha4w1wno1w076qwb.png)
```
[the remark this answers is not in the document]
Comment (muratkeremozcan) [reply sections/tools/visual-regression-testing.md]: moved all assets to images folder

**Review by NoriSte — APPROVED** [review]

**Pull request merged**